docs/design/proposals/horizontal-scaling.md

```diff
@@ -0,0 +1,223 @@
+# Horizontal scaling in KRO
+
+This proposal complements [KREP-002] and defines:
+- An interface to control `ResourceGraphDefinition` (RGD) instance scale using the Kubernetes “scale subresource”.
+- An opt-in mechanism to replicate RGD instances and declaratively control rollout across replicas.
+
+It is inspired by Kubernetes Set controllers (for example, `ReplicaSet`) to achieve safe, observable, and controllable rollouts.
+
+## Goals
+
+- Integrate KRO with the Kubernetes horizontal scaling ecosystem (for example, `kubectl scale`, [HPA], and [keda]) via the [scale-subresources].
+- Keep a single RGD instance as a “unit of deployment”; KRO’s graph resolution ensures correct ordering and propagation within that unit.
+- Provide an optional “Set” abstraction to replicate RGD instances and control rollout across replicas.
+
+## Non-goals
+
+- Implement all possible scaling or rollout strategies.
+- Replace or reimplement HPA/VPA behavior.
+- Manage progressive rollout within a single RGD instance. The Set abstraction controls rollout across replicas only.
+
+## Background and problem statement
+
+In [KREP-002] we introduced looping/collections. Collections may be required for:
+- Different environments (for example, private vs. public ingresses).
+- Different customers (duplicated stacks, different hostnames).
+- Horizontal scale based on load (request rate, queue depth, number of replicas, and so on).
+
+Required replica counts can vary rapidly. Kubernetes already provides a rich ecosystem (`kubectl scale`, [HPA], [keda]). KRO should integrate with this ecosystem.
+
+Replicas can be:
+- A single resource (for example, a Pod or a Deployment), or
+- A set of resources that together form a functional unit (for example, a [cluster-api] [capa-cluster]).
+
+When a replica is a set of resources, we need controlled rollout to limit blast radius if a new version/configuration causes issues.
+
+## Design overview
+
+This proposal proceeds in two phases:
+
+1) Scale subresource on RGD (integration with the Kubernetes scaling ecosystem).
+2) RGD Set controller (replicate RGD instances and control rollout across replicas).
+
+### Phase 1: Scale subresource on RGD
+
+Users add explicit schema annotations on RGD fields to expose the Scale subresource:
+- `| scale` on integer fields to represent desired/current replicas.
+- `| scaleSelector` on a status string field that holds a label selector for Pods (used by HPA/VPA).
+
+Validation rules:
+- `| scale` applies to integers only.
+- `| scale` must be present on both spec and status (or on neither).
+- `| scaleSelector` applies to a string attribute in status. KRO does not validate selector correctness nor enforce that it matches Pods.
+
+Example (minimum to expose Scale):
+
+```yaml
+apiVersion: kro.run/v1alpha1
+kind: ResourceGraphDefinition
+schema:
+  spec:
+    desiredReplicas: "integer | scale"
+  status:
+    replicas: "integer | scale"
+    podSelector: "string | scaleSelector"
```

Comment:
Noticing that a podSelector is not a string, it's a map[string]string. How would you model this?